substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSections.java

```diff
@@ -0,0 +1,2344 @@
+        // install details of heap elements
+        // String name = debugDataInfo.toString();
+        // }
+    }
+
+    public ClassEntry ensureClassEntry(Range range) {
+        String className = range.getClassName();
+        // see if we already have an entry
+        ClassEntry classEntry = primaryClassesIndex.get(className);
+        if (classEntry == null) {
+            // create and index the entry associating it with the right file
+            FileEntry fileEntry = ensureFileEntry(range);
+            classEntry = new ClassEntry(className, fileEntry);
+            primaryClasses.add(classEntry);
+            primaryClassesIndex.put(className, classEntry);
+        }
+        assert classEntry.getClassName().equals(className);
+        return classEntry;
+    }
+
+    public FileEntry ensureFileEntry(Range range) {
+        String fileName = range.getFileName();
+        // ensure we have an entry
+        FileEntry fileEntry = filesIndex.get(fileName);
+        if (fileEntry == null) {
+            DirEntry dirEntry = ensureDirEntry(fileName);
+            String baseName = (dirEntry == null ? fileName : fileName.substring(dirEntry.getPath().length() + 1));
+            fileEntry = new FileEntry(stringTable.uniqueDebugString(fileName),
+                            stringTable.uniqueString(baseName),
+                            dirEntry);
+            files.add(fileEntry);
+            filesIndex.put(fileName, fileEntry);
+            // if this is a primary entry then add it to the primary list
+            if (range.isPrimary()) {
+                primaryFiles.add(fileEntry);
+            } else {
+                Range primaryRange = range.getPrimary();
+                FileEntry primaryEntry = filesIndex.get(primaryRange.getFileName());
+                assert primaryEntry != null;
+            }
+        }
+        return fileEntry;
+    }
+
+    public void addRange(Range primaryRange, List<DebugFrameSizeChange> frameSizeInfos, int frameSize) {
+        assert primaryRange.isPrimary();
+        ClassEntry classEntry = ensureClassEntry(primaryRange);
+        PrimaryEntry entry = classEntry.addPrimary(primaryRange, frameSizeInfos, frameSize);
+        if (entry != null) {
+            // track the entry for this range in address order
+            primaryEntries.add(entry);
+        }
+    }
+
+    public void addSubRange(Range primaryRange, Range subrange) {
+        assert primaryRange.isPrimary();
+        assert !subrange.isPrimary();
+        String className = primaryRange.getClassName();
+        ClassEntry classEntry = primaryClassesIndex.get(className);
+        FileEntry subrangeEntry = ensureFileEntry(subrange);
+        // the primary range should already have been seen
+        // and associated with a primary class entry
+        assert classEntry.primaryIndexFor(primaryRange) != null;
+        classEntry.addSubRange(subrange, subrangeEntry);
+    }
+
+    private DirEntry ensureDirEntry(String file) {
```

Comment:
Use Path here an you get Platform portability for free.

Reply:
I have pushed some changes which address this problem. I modified the interface between the native image writer and the object file to pass a String for file name and Path for directory. The object file writer also employs Path internally to store both dir names and full (dir-qualfiied) file names. This avoids issues any need to handle path separators.

I'm not sure yet that this is all done though. In particular, the info in the hosted class and source node positions does not contain any information about original source file locations. So, the native image writer is still populating Path instances with the component elements of the hosted class's package name or the inlined class's qualified name, respectively.

Reply:
Thsi has been reworked. See other comments.